[3-4단계 방탈출 예약 대기] 드라고(이용화) 미션 제출합니다.

Firedrago95 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
 [방탈출 예약대기 프롤로그](https://prolog.techcourse.co.kr/studylogs/4321)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
디미터의 법칙등 객체지향 생활체조 원칙을 지키지 못한 부분들이 있고, 엔티티 설계가 객체지향적인 방식인지 의문이 있기 때문입니다.

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
## 이번 미션의 우선순위 - 도메인 규칙을 담은 엔티티 설계
이번 미션에서 집중한 세부 목표는 엔티티 설계입니다. 예약 대기는 예약이 존재해야만 생성될 수 있고, 기존에 존재하는 예약에 대해서 생성된다고 생각했기 때문에 예약 대기가 예약에 의존하도록 설계했습니다. 코드 레벨에서의 유효성 검증 누락을 방지할 수 있고, 데이터 무결성도 보장할 수 있다는 장점이 있습니다. 다만 예약이 삭제될 경우, 해당 예약을 참조하는 모든 예약 대기가 영향을 받게 되는 문제가 있었습니다. 그래서 예약을 삭제하는 대신 해당 예약의 예약자를 변경하고 우선순위의 예약대기를 삭제하는 방식을 사용하였습니다.

## 연관관계 편의 메서드
```java
private void setMember(Member member) {
        this.member = member;
        member.getReservations().add(this);
    }
```
기존에는 `Reservation` 객체가 생성될 때 `Member`를 생성자 혹은 `setMember()`를 통해 설정하고, 동시에`member.getReservations().add(this)`로 리스트에 추가하는 방식으로 연관관계를 맺고 있었습니다. `Member`에서는 별도의 `addReservation()` 메서드 없이 단순히 `getReservations()`를 통해 외부에서 직접 리스트에 접근하고 추가할 수 있는 구조였습니다.
양방향 관계를 설정할대, **예약이 생성되는 경우에는 멤버에 반드시 예약이 추가되지만, 멤버가 추가되는 경우 예약이 반드시 추가되지는 않으므로 양방향 관계를 `Reservation`에서만 관리해야 한다고 생각했습니다.** 이 경우, `Reservation`과 `Member` 사이의 연관관계가 한쪽만 세팅되거나, 두 객체의 상태가 불일치하는 문제가 발생할 수 있습니다.

```java
Member.addReservation(Reservation reservation) {
    this.reservations.add(reservation)
    if (!reservation.hasSameMember(this)) {
        reservation.setMember(this)
    }
}

Reservation.setMember(Member member) {
    this.member = member
    if (!member.contains(this)) {
        member.addReservation(member)
    }
}
```
이를 해결하기 위해 이번에 연관관계 편의 메서드를 양쪽 모두에 정의하는 대신, 순환참조 문제를 방지하기 위해서 조건문을 추가할 수 있습니다. 즉, `Member.addReservation()`에서는 `reservation.setMember(this)`를 호출하기 전에 `reservation.hasSameMember(this)` 조건을 확인하고, 반대로 `Reservation.setMember()`에서는 `member.contains(this)` 여부를 체크한 후 `member.addReservation(this)`를 호출합니다. 이렇게 하면 순환참조 없이 양방향 관계를 유지할 수 있습니다.

이와 별개로 이번 요구사항을 구현하면서 양방향 연관관계를 단방향 관계로 수정했습니다. 회원이 자신의 예약 목록을 조회해야 하는 요구사항이 있었기 때문에 Member → Reservation으로의 일대다 양방향 연관관계를 설정했습니다. 이후 예약 외에도 예약 대기 상태까지 함께 조회해야 하는 요구사항이 추가되면서, Member가 관리해야 하는 연관관계가 두 개로 늘어나는 상황이 되었습니다. 각 연관관계에 대해 동일한 방식의 편의 메서드를 별도로 관리해야 하고, 잘못 관리될 경우 상태 불일치 위험이 커질 수 있다고 판단하여 단방향 관계로 변경하였습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Firedrago95 committed:
feat: 예약 대기 클라이언트 코드 추가

- '예약 대기' 버튼 및 관련 이벤트 핸들러 추가
- 예약 상태에 따라 '예약 대기' 버튼 활성화/비활성화 처리 로직 구현
- '예약 대기' API 호출 함수(onWaitButtonClick) 추가
- reservation.html 템플릿에 '예약 대기' 버튼 추가

Firedrago95 committed:
docs: 예약 대기 기능 문서 추가

- README.md에 '예약 대기' 관련 API 명세 추가
- '/reservations/waiting' 엔드포인트 및 동작 조건 설명 포함

Firedrago95 committed:
feat: 예약대기 요청 기능 추가

- 예약대기 생성 로직 및 관련 유효성 검증 추가
- Waiting, WaitingRepository, JpaWaitingRepository 등 새로운 도메인 및 인터페이스 구현
- '/reservations/waiting' API 엔드포인트 추가
- 중복 예약대기, 예약 소유자 예약대기 방지 로직 구현
- 관련 테스트 케이스 추가 및 init, data SQL 업데이트

Firedrago95 committed:
fix: 예약 데이터 전송 키 수정

- reservationData 객체에서 themeId, timeId로 키 이름 변경
- 기존 잘못된 키(theme, time)로 인해 발생 가능한 문제 수정

Firedrago95 committed:
feat: 멤버별 예약·대기 조회 기능 추가

- 멤버별 예약과 예약 대기를 함께 조회하는 기능 구현
- WaitingWithRank 클래스 추가하여 예약 대기 순위를 반환하도록 수정
- JpaWaitingRepository와 관련 인터페이스에 멤버 ID로 예약 대기를 조회하는 메서드 추가
- 테스트 코드 수정 및 새로운 테스트 케이스 추가
- README에서 API 명세 업데이트

Firedrago95 committed:
feat: 예약 대기 삭제 기능 추가

- 예약 대기를 특정 id를 통해 삭제할 수 있는 API 구현
- 존재하지 않는 예약 대기 삭제 시 예외 처리 추가
- 프론트엔드에서 예약 대기 삭제 요청을 위한 엔드포인트 연동
- 예약 대기 관련 리포지토리에 조회 및 삭제 메서드 추가

Firedrago95 committed:
feat: 예약 대기 관리자 페이지 추가

- "/admin/waiting" URL로 예약 대기 관리 페이지 접근 가능
- 예약 대기 데이터를 조회 및 렌더링하는 JS 추가
- 예약 대기 관리 페이지의 HTML 템플릿 구성
- README에 해당 기능 추가 완료 사항 업데이트

Firedrago95 committed:
feat: 관리자용 예약 대기 목록 조회 기능 추가

- "/admin/waitings" GET endpoint를 통해 모든 예약 대기를 조회 가능하도록 구현
- ReservationService에 findAllWaitings 메서드 추가
- AdminReservationController에 관리자 인증(@LoginAdmin) 관련 로직 추가
- ReservationResponse에 Waiting 데이터를 변환하는 정적 메서드 추가
- JpaWaitingRepository에 findAll 메서드 구현
- 관련 테스트 코드 추가

Firedrago95 committed:
test: 예약 관련 테스트 코드 가독성 개선 및 검증 로직 추가

- 불필요한 주석 제거로 테스트 코드 간결화
- `can_find_all_waitings` 테스트에 추가 검증 로직 추가
- `Theme` 클래스 import 선언 추가

Firedrago95 committed:
feat: 예약 대기 취소 기능 추가

- "/admin/waiting/{id}" delete API를 통해 예약 대기를 취소할 수 있는 기능 구현
- Frontend에서 예약 대기 거절 버튼 활성화 및 API 호출 로직 추가
- Readme에 해당 기능 업데이트

Firedrago95 committed:
feat: 예약 삭제 시 대기 예약 처리 기능 추가

- 예약 삭제 시 대기 목록이 있을 경우 첫 번째 대기 예약자로 재지정
- 예약 삭제 로직에 대기 목록 처리 로직 추가 및 테스트 코드 작성
- Member와 Reservation 간 양방향 의존성 제거
- JPA 리포지토리에 대기 목록 및 사용자별 예약 조회 메소드 추가

Firedrago95 commented on src/main/resources/templates/admin/waiting.html:
```diff
@@ -0,0 +1,69 @@
+
+<div class="content-container">
+    <h2 class="content-container-title">예약 대기 관리 페이지</h2>
+    <div class="table-container"/>
```
html, js, css 코드에 대한 리뷰 중단 요청 미션에서 중요한 부분이 아님

Firedrago95 commented on src/main/resources/static/js/waiting.js:
```diff
@@ -0,0 +1,89 @@
+    }).then(() => location.reload());
+}
+
+function deny(event) {
+    const row = event.target.closest('tr');
+    const id = row.cells[0].textContent;
+
+    /*
+    TODO: [4단계] 예약 대기 목록 관리 기능
+          예약 대기 거절 API 호출
+     */
+    const endpoint = `/admin/waiting/${id}`;
+    return fetch(endpoint, {
+        method: 'Delete'
+    }).then(response => {
+        if (response.status === 200) return;
+        throw new Error('Delete failed');
+    }).then(() => location.reload());
+}
```
html, js, css 코드에 대한 리뷰 중단 요청 미션에서 중요한 부분이 아님

Firedrago95 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -93,10 +128,27 @@ public List<ReservationResponse> getReservations() {
+        Reservation reservation = reservationRepository.findById(id)
+            .orElseThrow(() -> new ReservationException("예약을 찾을 수 없습니다."));
+
+        List<Waiting> waitings = waitingRepository.findByReservationId(id);
+        if (waitings.isEmpty()) {
+            reservationRepository.deleteById(id);
+            return;
+        }
+
+        Waiting firstWaiting = waitings.getFirst();
+        reservation.reAssignedTo(firstWaiting.getMember());
```
호출 쿼리에서 id 기반으로 'ORDER BY'로 정렬하여 조회하고 있습니다.

Firedrago95 committed:
fix: DTO 필드 타입 수정 및 잘못된 참조 변경

- WaitingResponse의 id 필드를 primitive 타입에서 Wrapper 타입으로 변경
- ReservationResponse에서 잘못된 객체 참조(waiting → reservation) 수정
- init.sql에 waiting 데이터 추가
- ReservationTime 생성자의 접근 제어자를 protected로 변경

Firedrago95 committed:
fix: 멤버 조회 시 예외 타입 변경

- 멤버를 찾을 수 없는 경우 사용되는 예외를 MemberNotFound로 수정
- 중복된 예외 타입 사용을 개선하여 코드 가독성과 유지보수성 향상

Firedrago95 committed:
test: JpaRepository 테스트 추가

- JpaWaitingRepository 테스트 클래스 추가
- JpaReservationRepository 테스트 클래스 추가
- JpaMemberRepository 테스트 클래스 추가
- JpaAdminRepository 테스트 클래스 추가
- JpaThemeRepository 테스트 클래스 추가
- JpaReservationTimeRepository 테스트 클래스 추가
- Admin 도메인에 정적 팩토리 메서드 추가

jamie9504 reviewed (changes requested):
안녕하세요, 드라고. 리뷰어 제이미입니다.

슬랙으로 대화 나눈 연관관계 편의 메서드에 대한 드라고의 정리를 PR 본문에 작성 후 재요청 주세요.

PR 본문이 미완성으로 보여, 코멘트 남기지 않고 바로 RC 보냅니다.

jamie9504 commented on readme.md:
```diff
@@ -1,26 +1,37 @@
 ## 기능 구현 목록
```
양방향 연관관계에 대해 정리해주셨네요.
양방향으로 연관관계를 갖게 되면 관리해야 하는 포인트가 늘어나기에 꼭 필요한 경우에만 양방향을 쓰는 습관을 들이면 좋습니다.

jamie9504 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -63,7 +62,15 @@ public String getPassword() {
+    public boolean equals(Object o) {
+        if (o == null || getClass() != o.getClass()) return false;
+        Member member = (Member) o;
+        return Objects.equals(id, member.id) && Objects.equals(name, member.name) && Objects.equals(email, member.email) && Objects.equals(password, member.password);
```
DB 컬럼의 경우 PK가 같으면 같은 컬럼으로 구분하는데, Equals & HashCode를 구현한 기준은 무엇인가요?

Firedrago95 replied:
엔티티의 동등성을 비교할때 모든 필드를 사용할 경우, 필드의 값이 변경되면 다른 객체로 인식되어 문제가 생길 수 있습니다. 일반적으로 id (식별자)를 기준으로 동등성을 비교한다는 것을 알게되었습니다. 다만 id는 엔티티가 영속상태가 되었을때 부여되기 때문에, 영속상태가 아닌 엔티티의 동등성 비교가 필요하다면 동등성을 보장하기 위한 별도의 필드를 설정해야 한다는 사실 역시 알게되었습니다. 현재 Member는 영속상태에서만 비교가 진행되므로 id를 사용하여 equals & HashCode를 구현하도록 수정하였습니다.

[fix: equals/hashCode 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/3fb449689600aac0b6ab904ed8f4e7c120f69c9d)

jamie9504 replied:
만약 ResponseBody로 일괄 입력을 받고, 중복 제거를 한 후 저장하는 로직이 추가된다면 어떻게 될까요?

Firedrago95 replied:
Member의 여러 필드는 유저 정보와 관련되어 있으며, 대부분 수정 가능한 값들이라고 생각합니다. 이러한 변화 가능한 필드들이 equals와 hashCode에 포함될 경우, 필드 값이 변경되었을 때 동일한 회원임에도 불구하고 다른 객체로 판단될 위험이 있습니다.
따라서 equals와 hashCode는 id만을 기준으로 구현하고, 아직 영속 상태가 아닌 Member 객체들 간의 비교가 필요한 경우에는 서비스 계층에서 특정 필드 값을 활용해 중복을 직접 판별하는 방식이 더 안전한 방식이라고 생각합니다.

jamie9504 commented on src/main/java/roomescape/member/presentation/dto/MyReservationResponse.java:
```diff
@@ -2,16 +2,38 @@
 public record MyReservationResponse(Long reservationId, String theme, LocalDate date, LocalTime time, String status) {
+
+    public static final String RESERVED = "예약";
+    public static final String WAITING = "%d번째 예약대기";
```
public이 적합한가요?

---

코드 전반적으로 마지막 줄 개행 추가해주세요. 기본적인 부분 다 누락 없이 체크해주세요.

Firedrago95 replied:
protected로 수정하였습니다. 테스트 클래스에 개행을 추가하였습니다.

[refactor: 예약 상태 상수 접근 제어자 변경](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/c21a126548748cd186a17f3f4dfc0c1dd49ee8c7)
[style: 개행 추가](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/5bf51200573439a0373e4ef1fcc7dc75d482a039)

jamie9504 commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -88,6 +83,14 @@ public boolean isSameTime(final ReservationTime time) {
+        return this.member.equals(member);
+    }
+
+    public void reAssignedTo(Member member) {
```
단순한 세터 메서드로 보이는데, 이렇게 이름을 정의한 이유가 있나요?

Firedrago95 replied:
양방향 연관관계를 유지하기 위한 편의메서드에서 양방향 연관관계가 제거되면서 이름을 변경하지 않았던 실수가 있었습니다. setter() 메서드로 이름을 수정하였습니다.

[refactor: 예약 재배정 로직 메서드명 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/eec2e711b06fbe60cc50297da4449f0523fe3def)

jamie9504 commented on src/main/java/roomescape/reservation/domain/WaitingWithRank.java:
```diff
@@ -0,0 +1,19 @@
+package roomescape.reservation.domain;
+
+public class WaitingWithRank {
```
Domain인가요?

Firedrago95 replied:
`WaitingWithRank`는 직접적으로 db 컬럼과 연관이 있는 엔티티가 아닌점, 컬럼에서 순위를 찾아 맵핑하는 dto의 역할을 한다고 생각하여 infrastructure.dto 패키지로 위치를 변경하였습니다.

[refactor: WaitingWithRank 클래스 위치 변경 및 패키지 구조 정리](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/6984587f7c11b7bb9cdb926a9a57b7d2193bd08b)

jamie9504 commented on src/main/java/roomescape/reservation/presentation/AdminReservationController.java:
```diff
@@ -47,4 +45,15 @@ public ResponseEntity<List<ReservationResponse>> reservationFilter(@ModelAttribu
+    @DeleteMapping("/admin/waiting/{id}")
+    public ResponseEntity<Void> deleteWaiting(@LoginAdmin LoginAdminInfo adminInfo, @PathVariable("id") Long waitingId) {
+        reservationService.deleteWaiting(waitingId);
+        return ResponseEntity.ok().build();
```
1. `/admin/waiting`과 `/admin/waitings`가 혼용되고 있네요.
이렇게 사용한 이유는 무엇인가요?

2. 삭제 후 별도의 메시지를 내리지 않을 때 적합한 HttpStatusCode는 무엇이라고 생각하시나요?
이번 구현에 여러 방법이 혼용되어 있네요.

Firedrago95 replied:
1. 처음에는 delete 메서드는 하나의 자원을 삭제하기 때문에 단수형(`/waiting`), get 메서드는 여러 자원을 조회하기 때문에 복수형(`/waitings`)으로 사용하는 것이 명확하다고 생각했습니다. 하지만 REST API 설계에서는 URL이 '동작'이 아닌 '리소스'를 지칭한다는 점을 알게되었습니다. 그래서 복수/단수를 혼용하는 방식이 오히려 혼동을 줄 수 있다는 걸 알게 되었습니다. 특히 동일한 리소스를 다루는 API에서 단수/복수가 혼용되면 일관성 측면에서 문제가 될 수 있다고 느꼈고, 실제 테이블(리소스)을 지칭하는 복수형으로 통일하는 것이 바람직하다고 판단했습니다. URL은 복수형을 사용하도록 수정하였습니다.
[refactor: API 경로 및 HTTP 메서드 일관성 개선](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/af1733aa1722eb2bc535dae4137140d4ab87a8a4)

2. 삭제 요청이 성공한 경우 `204 NoContent` 상태를 반환하는 것이 적합하다고 생각합니다. 삭제 요청이 성공할 경우 `204 No Content` 상태를 반환하도록 수정하였습니다.
[fix: DELETE 요청 시 응답 코드를 204로 변경](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/4751864c419532e72f2c83d18c70a29bb9c70873)

jamie9504 commented on src/main/java/roomescape/reservation/domain/ReservationRepository.java:
```diff
@@ -18,9 +18,13 @@ public interface ReservationRepository {
 
     boolean existByThemeId(Long themeId);
+
+    Optional<Reservation> findBy(LocalDate date, Long aLong, Long aLong1);
```
여기에 더해서 findBy가 내려주고 있는 타입이 파라미터에 따라 다른데요.

`List<Reservation> findBy(Long memberId);`
`Optional<Reservation> findBy(LocalDate date, Long aLong, Long aLong1);`

직관성이 떨어져보입니다.

Firedrago95 replied:
List를 반환하는 경우 `findAllBy()`, Optional 단건 조회를 하는 경우 `findBy()` 사용하도록 리팩토링 하였습니다.

[refactor: findBy 메서드 명칭을 findAllBy로 변경](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/42d20b6183ea778d328fac8279380acd3d9d6edd)

jamie9504 commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -17,7 +17,7 @@ public class Theme {
     private String thumbnail;
 
-    protected Theme() {
+    public Theme() {
```
다시 Public으로 변경한 이유는 무엇인가요?

Firedrago95 replied:
기본생성자의 접근제어자를 다시 `protected`로 변경하였습니다.

[refactor: Theme 생성자의 접근 제어자를 protected로 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/c23ec66ae610504e489a167519e76953f920ef23)

jamie9504 commented on src/test/java/roomescape/admin/infrastructure/JpaAdminRepositoryTest.java:
```diff
@@ -0,0 +1,38 @@
+    private JpaAdminRepository jpaAdminRepository;
+
+    @Test
+    void 이메일로_Admin을_조회할_수_있다() {
```
테스트를 지금과 같이 한다면, 어떤 부분을 보장할 수 있고 어떤 부분을 보장할 수 없을까요?
이 테스트로 보장받고 싶은 부분은 무엇일까요?

Firedrago95 replied:
RepositoryTest를 통해 Spring Data JPA가 생성한 구현체가 의도한 작업을 잘 수행하는지 확인하고자 했습니다. 이를 위해 `@DataJpaTest` 어노테이션을 사용하여 실제 JPA 구현체를 주입받아 테스트를 구성하였습니다. 현재 테스트 방식이 보장하는 것은 JPA 쿼리 메서드(findByEmail, existsByEmail 등)가 의도한 조건에 따라 정확하게 동작하는지 여부입니다. 반면, 보장하지 못하는 부분은 테스트에 필요한 초기 데이터가 data.sql에 정의되어 있기 때문에, 해당 데이터를 테스트 코드만 보고는 직관적으로 파악하기 어렵다는 점입니다. 또한, 테스트가 data.sql에 의존하고 있기 때문에 파일이 변경되면 테스트가 예기치 않게 깨질 수 있는 위험도 존재합니다. 그럼에도 불구하고, 이번 테스트에서는 'Spring Data JPA 구현체의 쿼리 동작 확인'이 주 목적이었고, data.sql을 통한 초기 데이터 구성은 이 목적에 크게 방해되지 않는다고 판단하여 현재 방식으로 작성하였습니다.

jamie9504 replied:
간단한 동작이라면 save를 하는 것이 오히려 가독성은 더 좋을 수도 있겠네요.

언제 직접 save를 호출하고 언제 data.sql을 사용하는지 기준이 있을까요?

Firedrago95 replied:
data.sql은 데이터의 의존관계가 복잡하고, 테스트에 필요한 데이터 값이 많고, 여러 테스트에서 공통으로 필요한 데이터값이 필요할때 사용할 수 있습니다. save()의 경우 데이터의 의존관계가 비교적 간단하고, 명시적으로 데이터를 보여줘야 할때 사용할 수 있습니다.

jamie9504 reviewed (changes requested):
안녕하세요, 드라고. 리뷰어 제이미입니다.

간단한 코멘트들만 남겨두었습니다.

Firedrago95 committed:
style: 개행 추가

- 모든 테스트 파일의 마지막에 newline 추가
- 코드 포맷 통일성을 유지하기 위한 수정

Firedrago95 committed:
refactor: 예약 상태 상수 접근 제어자 변경

- MyReservationResponse의 예약 상태 상수를 public에서 private으로 변경
- 불필요한 외부 접근을 제한하고 캡슐화 강화

Firedrago95 committed:
refactor: 예약 재배정 로직 메서드명 수정

- `reAssignedTo` 메서드명을 `setMember`로 변경하여 직관성을 향상
- 호출부에서도 변경된 메서드명으로 수정

Firedrago95 committed:
refactor: Theme 생성자의 접근 제어자를 protected로 수정

Firedrago95 committed:
refactor: WaitingWithRank 클래스 위치 변경 및 패키지 구조 정리

- WaitingWithRank를 domain 패키지에서 infrastructure.dto 패키지로 이동
- 관련된 클래스 및 테스트 파일의 import 경로 수정
- 불필요한 import 제거로 코드 정리

Firedrago95 committed:
refactor: API 경로 및 HTTP 메서드 일관성 개선

- 예약 대기 API 경로 '/admin/waiting'을 '/admin/waitings'로 변경
- 내 예약 조회 API 경로 '/reservations-mine'을 '/reservations/me'로 변경
- 회원 예약 검색 API 경로 '/user-reservation'을 '/admin/reservations'로 변경
- HTTP 메서드 'Delete'를 'DELETE'로 수정

Firedrago95 committed:
fix: DELETE 요청 시 응답 코드를 204로 변경

- JavaScript 및 Controller의 DELETE 응답 코드를 200에서 204로 수정
- HTTP DELETE의 표준에 맞게 수정하여 일관성 유지

Firedrago95 committed:
refactor: findBy 메서드 명칭을 findAllBy로 변경

- 일관성 있는 메서드명을 위해 findBy를 findAllBy로 수정
- 관련 서비스 및 레포지토리에 변경사항 반영

Firedrago95 committed:
fix: equals/hashCode 수정

- Member 클래스의 equals 및 hashCode 메서드에서 id만 비교하도록 변경

jamie9504 commented on src/main/java/roomescape/reservation/infrastructure/JpaWaitingRepository.java:
```diff
@@ -0,0 +1,37 @@
+    WHERE w.reservation.id = :reservationId
+    ORDER BY w.id ASC
+    """)
+    List<Waiting> findByReservationId(@Param("reservationId") Long reservationId);
```
ORDER BY는 jpql로 생성해야만 할까요?

Firedrago95 replied:
`findByReservationIdOrderById()` 메서드명을 통해 spring-data-jpa 기능으로 ORDER BY 절을 자동 생성하도록 변경하였습니다.

[refactor: 대기자 조회 메서드 이름 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/aebd18f00177a3d7ff7d0840569998bb3dcfbeb5)

jamie9504 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -70,6 +75,34 @@ public ReservationResponse createReservation(final ReservationRequest request, f
+
+        validateNotReservationOwner(reservation, member);
+        validateCanReserveDateTime(reservation, dateTime.now());
+        validateDuplicateWaiting(reservation, member);
+
+        Waiting waiting = waitingRepository.save(Waiting.createWithoutId(reservation, member));
```
있는지 조회 후 없다면 등록하는 것과

등록 후 중복건이 있어서 (DB Unique 조건 이용 등) 실패한다면 예외를 던지는 건 어떠한 차이가 있을까요?

Firedrago95 replied:
제이미가 질문주셔서 '중복 검증과 동시성' 문제에 대해 처음 알게되었습니다. 검증로직과 등록 로직이 따로 호출될때, A,B 두 요청이 거의 동시에 검증 로직을 통과하게 되고 같은 예약 정보가 등록될 위험이 있습니다. unique 조건을 설정하면, db가 동시에 입력 요청이 들어와도 하나의 요청만 통과하고 예외를 발생시켜 동시성 문제를 예방할 수 있다는 것을 알게되었습니다. 만약 동시성이 고려되어야하는 환경이라면 밑과 같은 설정을 엔티티에 추가하고, try - catch를 통해 발생되는 예외를 처리해줘야 합니다.
```java
@Table(name = "waiting", 
uniqueConstraints = {
    name = "uk_reservation_theme_date_time"
    @UniqueConstraint(columnNames = {"theme_id", "date", "time_id"})
```
```java
@Transactional
public ReservationResponse createReservation(final ReservationRequest request, final Long memberId) {
    try {
    // 로직
    } catch (DataIntegrityViolationException e) {
        if (e.getMessage().contains("uk_reservation_theme_date_time")) {
            throw new BusinessException("해당 테마의 해당 날짜와 시간에 이미 예약이 존재합니다.");
        }
        throw e;
    }
```

jamie9504 commented on src/main/java/roomescape/reservation/infrastructure/JpaWaitingRepositoryAdapter.java:
```diff
@@ -0,0 +1,54 @@
+        this.jpaWaitingRepository = jpaWaitingRepository;
+    }
+
+    @Override
+    public boolean exists(Long reservationId, Long memberId) {
+        return jpaWaitingRepository.existsByReservationIdAndMemberId(reservationId, memberId);
+    }
+
+    @Override
+    public Waiting save(Waiting waiting) {
+        return jpaWaitingRepository.save(waiting);
+    }
+
+    @Override
+    public List<WaitingWithRank> findByMemberId(Long memberId) {
+        return jpaWaitingRepository.findByMemberId(memberId);
+    }
```
exists와 find는 네이밍 기준이 다르네요.
하나는 파라미터를 명시해주고 있어요.

Firedrago95 replied:
파라미터를 명시해주는 방향으로 메서드 이름 규칙을 통일하였습니다.

[refactor: 메서드 명명 규칙 수정 및 일관성 향상](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/f048a5bf12a53fa582b76d1b5148875d72a85b34)

jamie9504 commented on readme.md:
```diff
@@ -1,26 +1,37 @@
+  - [x] 동일한 사용자가 중복 예약 대기를 생성할 때
+- [x] "/reservations/waiting/{id}" delete 요청으로 예약 대기를 삭제 할 수 있다.
+- [x] "/admin/waitings" get 요청을 통해 관리자는 모든 예약 대기를 조회 할 수 있다.
+- [x] "/admin/waiting/{id}" delete 요청을 통해 관리자는 예약 대기를 취소 할 수 있다.
```
문서가 일부 최신화되지 않은거군요.

드라고는 ID로 단 건 조회할 때 URL을 어떻게 설계하는 것이 좋다고 생각하시나요?
생각한 방식으로 코드를 일관성있게 수정해보세요.

Firedrago95 replied:
저는 단건 조회의 경우라도 복수형을 사용하는 것이 좋다고 생각합니다. 단건조회든 여러건의 조회든 동일한 테이블 리소스에 요청을 보내는 것이기 때문입니다. 테이블은 자바의 컬렉션처럼 여러 데이터를 보관할 수 있으므로 복수형이 적절하다고 생각했습니다. 같은 이유로 페이지를 호출하는 URL의 경우에 테이블에 요청을 보내는 것이 아니라, 페이지를 호출하는 것이므로 단수형으로 표현하는 것이 적절하다고 생각했습니다. 
[docs: API 엔드포인트 경로 수정 반영](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/ccacdfb13d62d5e8f0ef01a067a664516b60fd1e)

jamie9504 commented on src/main/java/roomescape/reservation/domain/Waiting.java:
```diff
@@ -0,0 +1,43 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
LAZY 로딩을 선 적용하는 것과 후 적용하는 것은 어떠한 작업적 차이가 있을까요?

Firedrago95 replied:
기본 전략인 EAGER로 설정된 상태에서 컬렉션을 조회하여 순회하는 등의 상황이 있다면, 엄청난 양의 데이터가 한꺼번에 조회될 위험이 있습니다. 그런 상황을 방지하기 위해 일단 LAZY 지연로딩을 기본적으로 설정하고, 연관관계를 한번에 불러와야하는 상황이라면 join fetch등을 사용하여 이후에 최적화 할 수 있습니다. 즉 LAZY를 선적용하여, 즉시로딩을 통해 수많은 데이터가 한꺼번에 로드되는 상황을 막고 개발자가 통제할 수 있는 상황에서 쿼리를 최적화 할 수 있습니다.

jamie9504 commented on src/main/java/roomescape/theme/domain/ThemeRepository.java:
```diff
@@ -15,7 +12,7 @@ public interface ThemeRepository {
     void deleteById(Long id);
 
-    List<Theme> findPopularThemes(LocalDate start, LocalDate end);
+    List<Theme> findAllPopularThemes(LocalDate start, LocalDate end);
```
All을 사용하는 경우와 복수형을 사용하는 경우와 모두를 사용하는 것은 각각 어떠한 의미의 차이를 두고 있나요?

Firedrago95 replied:
'유명 테마를 찾는다' 는 의미는 비지니스 조건이므로 꼭 명시되어야 한다고 생각합니다. 그런데 복수형을 사용하고 있으므로 비지니스 조건이 아닌 단순 검색 이라면 all을 사용하고, 비지니스 조건으로 메서드를 표현할 수 있는 경우 all을 생략하도록 수정했습니다.

[refactor: 인기 테마 조회 메서드명 변경](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/16f7c9b0d1faac943ed90dbf5270b1674e08d0d8)

jamie9504 commented on src/test/java/roomescape/reservation/service/ReservationServiceTest.java:
```diff
@@ -61,25 +67,165 @@ void can_find_my_reservation() {
+            .isInstanceOf(ReservationException.class);
+    }
+
+    @DisplayName("지나간 날짜와 시간에 대한 예약대기를 생성할 수 없다.")
```
컨트롤러에 대한 테스트는 별도로 추가되지 않은 이유가 있나요?

Firedrago95 replied:
대부분의 컨트롤러 로직은 서비스의 로직을 단순 호출 하는 경우가 많고, 로그인 클래스의 경우 쿠키 생성 로직이 테스트 할 필요가 있다고 생각하여 LoginController 테스트를 추가하였습니다.

[test: 로그인 컨트롤러 테스트 추가](https://github.com/woowacourse/spring-roomescape-waiting/pull/272/commits/ebe2cb8fef06917bfef8bd4506e0672c74fc2238)

jamie9504 reviewed (changes requested):
안녕하세요, 드라고. 리뷰어 제이미입니다.

간단한 코멘트 남겨두었습니다.

다음 요청시 특이사항 없으면 Approve & Merge 예정입니다.

Firedrago95 committed:
refactor: 대기자 조회 메서드 이름 수정

- 메서드명을 findByReservationIdOrderById로 변경하여 명확성 증가
- Jpa 쿼리 대신 Spring Data JPA 기능을 사용해 정렬 로직 단순화
- 관련 테스트 코드에서 변경된 메서드 호출로 수정

Firedrago95 committed:
docs: API 엔드포인트 경로 수정 반영

- 특정 사용자 예약 조회 및 예약 대기 관련 API 경로 수정 사항 반영
- 관리자 예약 대기 관리 API 경로 변경 내용 업데이트

Firedrago95 committed:
refactor: 메서드 명명 규칙 수정 및 일관성 향상

- ReservationRepository 및 WaitingRepository 관련 메서드 명명을 더 명확하고 직관적으로 변경
- 기존 메서드(findAllBy 등) 이름을 'findAllByDateAndThemeId' 등으로 수정하여 의미 전달 강화
- 여러 서비스 클래스에서 변경된 메서드명을 참조하도록 수정

Firedrago95 committed:
refactor: 인기 테마 조회 메서드명 변경

- 메서드명을 findAllPopularThemes에서 findPopularThemes로 변경
- 메서드 명칭 수정으로 가독성과 코드 일관성 개선

Firedrago95 committed:
test: 로그인 컨트롤러 테스트 추가

- 회원 로그인 성공 테스트 추가
- 관리자 로그인 성공 테스트 추가

jamie9504 reviewed (approved):
안녕하세요, 드라고. 리뷰어 제이미입니다.

마감기한 이후 요청이라 별도의 코멘트 없이 Approve & Merge 합니다.

2레벨 마지막 미션도 화이팅입니다.
